pss/crypto/crypto.go

```diff
@@ -0,0 +1,656 @@
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, nil, err
+	}
+	decrypted, err = aesgcm.Open(nil, salt, rawBytes[:len(rawBytes)-aesNonceLength], nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	return
+}
+
+// encryptAsymmetric encrypts a message with a public key.
+func (crypto *defaultCryptoBackend) encryptAsymmetric(rawBytes []byte, key *ecdsa.PublicKey) ([]byte, error) {
+	if !validatePublicKey(key) {
+		return nil, errInvalidPubkey
+	}
+	encrypted, err := ecies.Encrypt(crand.Reader, crypto.importECDSAPublic(key), rawBytes, nil, nil)
+	if err == nil {
+		return encrypted, nil
+	}
+	return nil, err
+}
+
+func (crypto *defaultCryptoBackend) decryptAsymmetric(rawBytes []byte, key *ecdsa.PrivateKey) ([]byte, error) {
+	return ecies.ImportECDSA(key).Decrypt(rawBytes, nil, nil)
+}
+
+func (crypto *defaultCryptoBackend) encryptSymmetric(rawBytes []byte, key []byte) ([]byte, error) {
+	if !validateDataIntegrity(key, aesKeyLength) {
+		return nil, errInvalidSymkey
+	}
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+	salt, err := crypto.generateSecureRandomData(aesNonceLength) // never use more than 2^32 random nonces with a given key
+	if err != nil {
+		return nil, err
+	}
+	encrypted := aesgcm.Seal(nil, salt, rawBytes, nil)
+	encBytes := append(encrypted, salt...)
+	return encBytes, nil
+}
+
+// signHash calculates an ECDSA signature.
+func (crypto *defaultCryptoBackend) signHash(hash []byte, prv *ecdsa.PrivateKey) (sig []byte, err error) {
+	return ethCrypto.Sign(hash, prv)
+}
+
+// Obtain public key from the signed message and the signature
+func (crypto *defaultCryptoBackend) sigToPub(signed, sig []byte) (*ecdsa.PublicKey, error) {
+	hash := crypto.keccak256(signed)
+	return ethCrypto.SigToPub(hash, sig)
+}
+
+// GenerateRandomID generates a random string, which is then returned to be used as a key id
+func (crypto *defaultCryptoBackend) generateRandomID() (id string, err error) {
+	buf, err := crypto.generateSecureRandomData(keyIDSize)
+	if err != nil {
+		return "", err
+	}
+	if !validateDataIntegrity(buf, keyIDSize) {
+		return "", fmt.Errorf("error in generateRandomID: crypto/rand failed to generate random data")
+	}
+	id = common.Bytes2Hex(buf)
+	return id, err
+}
+
+// keccak256 calculates and returns the keccak256 hash of the input data.
+func (crypto *defaultCryptoBackend) keccak256(data ...[]byte) []byte {
+	return ethCrypto.Keccak256(data...)
+}
+
+// generateSecureRandomData generates random data where extra security is required.
+// The purpose of this function is to prevent some bugs in software or in hardware
+// from delivering not-very-random data. This is especially useful for AES nonce,
+// where true randomness does not really matter, but it is very important to have
+// a unique nonce for every message.
+func (crypto *defaultCryptoBackend) generateSecureRandomData(length int) ([]byte, error) {
```

Comment:
are we sure we want to reinvent the wheel here?
Surely we should expect reliable random data from crand.Read

Reply:
Just copied from Whisper. I don't have knowledge of the mentioned "bugs" in the comments. But maybe we can just remove it until we have more information.

Reply:
Actually, I just bought a USB hardware random number generator from the GNU shop. `crypto/rand` seems to suggest that the random source in linux should be selectable on the OS level, but it's less clear if that works well across platforms. If directly interfacing `crypto/rand` means you do not have a choice of random source, then I think we should keep the abstraction.